### PR TITLE
TDL-16006: Fix interruptible full table bookmarking strategy for bulk API

### PR DESCRIPTION
**Pull request by karanpanchal-crest**
# Description of change
 - Write unit test reproduce scenario where Full Table write bookmark with null key
 - Resolved Full Table bookmarking with null key state
 
# Risks

# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/unittests/test_salesforce_null_bookmark.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import logging
+from typing import Counter
+import unittest
+from unittest import mock
+# from tap_salesforce import main
```
Comment (savan-chovatiya) [inline tests/unittests/test_salesforce_null_bookmark.py:5]: Remove unused/commented import statements.

File: tests/unittests/test_salesforce_null_bookmark.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from tap_salesforce.sync import sync_records
+import json
+
+
+
+class TestNullBookmarkTesting(unittest.TestCase):
+    @mock.patch('tap_salesforce.salesforce.Salesforce.query', side_effect=lambda test1, test2: [])
+    def test_not_null_bookmark_for_incremental_stream(self, mocked_query):
+        """
+        To ensure that after resolving writebook mark logic not get "Null" key in state
```
Comment (savan-chovatiya) [inline tests/unittests/test_salesforce_null_bookmark.py:17]: Add detailed comment regarding what should be null in the state and which key-value pair is not getting emitted now.

Comment (karanpanchal-crest) [reply tests/unittests/test_salesforce_null_bookmark.py]: Added detailed comment for test function

File: tests/unittests/test_salesforce_null_bookmark.py (hunk 1 above)
Comment (karanpanchal-crest) [reply tests/unittests/test_salesforce_null_bookmark.py]: Remove and updated the code

File: tests/unittests/test_salesforce_null_bookmark.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import logging
+from typing import Counter
```
Comment (savan-chovatiya) [inline tests/unittests/test_salesforce_null_bookmark.py:2]: Remove these unused imports.

Comment (karanpanchal-crest) [reply tests/unittests/test_salesforce_null_bookmark.py:2]: Removed it and update the code

**Review by dbshah1212 — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

File: tests/unittests/test_salesforce_null_bookmark.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        
+        sf.pk_chunking = True
+        sf.instance_url = "https://cds-e-dev-ed.my.salesforce.com"
+        catalog_entry = {"stream": "OpportunityLineItem", "schema": {}, "metadata":[], "tap_stream_id": "OpportunityLineItem"}
+        state = {}
+        counter = metrics.record_counter('OpportunityLineItem')
+        sync_records(sf, catalog_entry, state, counter)
+        # write state function convert python dictionary to json string
+        state = json.dumps(state)
+        self.assertEqual(state, '{"bookmarks": {"OpportunityLineItem": {"version": null}}}', "Not get expected state value")
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/unittests/test_salesforce_null_bookmark.py:37]: @kspeer825 In the existing tap we are writing `version` as **null** for streams if the replication key is not found. [Reference](https://github.com/singer-io/tap-salesforce/blob/master/tap_salesforce/sync.py#L171)

File: tests/unittests/test_salesforce_null_bookmark.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        sf = Salesforce(
+            refresh_token="test",
+            sf_client_id="test",
+            sf_client_secret="test",
+            quota_percent_total=None,
+            quota_percent_per_run=None,
+            is_sandbox=None,
+            select_fields_by_default= False,
+            default_start_date='2019-02-04T12:15:00Z',
+            api_type="BULK")
```
Comment (luandy64) [inline tests/unittests/test_salesforce_null_bookmark.py:27]:
```suggestion
        sf = Salesforce(default_start_date='2019-02-04T12:15:00Z', api_type="BULK")
```

The tests still run without all of this extra code, so we shouldn't clutter the file with it to help increase the readability of the test

File: tests/unittests/test_salesforce_null_bookmark.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (luandy64) [reply tests/unittests/test_salesforce_null_bookmark.py:37]: I'd like to talk to the team about this tomorrow. I think the referenced code is a bug and we should be writing the version in the `ActivateVersion` message sent on Line 172 https://github.com/singer-io/tap-salesforce/blob/d89a18f4f4f95018cb508c8aea5522e74ea35b98/tap_salesforce/sync.py#L171-L174

**Review by luandy64 — CHANGES_REQUESTED** [review]

File: tests/unittests/test_salesforce_null_bookmark.py (hunk 5 above)
Comment (harshpatel4crest) [reply tests/unittests/test_salesforce_null_bookmark.py]: Updated the code to set minimal arguments.

**Review by luandy64 — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**